packages/core/core/generate-object/stream-object.ts

```diff
@@ -361,24 +359,27 @@ Response headers.
     let error: unknown | undefined;
 
     // pipe chunks through a transformation stream that extracts metadata:
-    let accumulatedText = '';
     let latestObject: DeepPartial<T> | undefined = undefined;
 
+    const streamingParser = new StreamingParser({ stream: true });
+
     this.originalStream = stream.pipeThrough(
       new TransformStream<string | ObjectStreamInputPart, ObjectStreamPart<T>>({
         async transform(chunk, controller): Promise<void> {
           // process partial text chunks
           if (typeof chunk === 'string') {
-            accumulatedText += chunk;
-
-            const currentObject = parsePartialJson(
-              accumulatedText,
-            ) as DeepPartial<T>;
-
-            if (!isDeepEqualData(latestObject, currentObject)) {
-              latestObject = currentObject;
-
-              controller.enqueue({ type: 'object', object: currentObject });
+            let { events } = streamingParser.parseIncremental(chunk);
+
+            // filter for events that change values
+            // @ts-expect-error TODO: fix type
+            events = events.filter(event => event.kind === 'partial' || (event.kind === 'complete' && typeof event.value !== 'object'))
+            if (events.length > 0) {
+              // @ts-expect-error
+              latestObject = structuredClone(streamingParser.root);
```

Comment:
The current `StreamingParser` does not expose its internal state, this `root` is the mutable partially parsed object.

This could be exposed as a `readonly DeepPartial<T>`.

Alternatively, an `applyEvents` function could maintain a parallel and safe external representation of the object. Without much more work, the object returned could be treated as a persistent data structure, creating new objects for each parent of a modified leaf.